src/main/java/cz/cvut/kbss/termit/service/term/OrphanedInverseTermRelationshipRemover.java

```diff
@@ -0,0 +1,75 @@
+package cz.cvut.kbss.termit.service.term;
+
+import cz.cvut.kbss.termit.dto.TermInfo;
+import cz.cvut.kbss.termit.model.Term;
+import cz.cvut.kbss.termit.persistence.dao.TermDao;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+
+@Component
+public class OrphanedInverseTermRelationshipRemover {
+
+    private static final Logger LOG = LoggerFactory.getLogger(OrphanedInverseTermRelationshipRemover.class);
+
+    private final TermDao termDao;
+
+    public OrphanedInverseTermRelationshipRemover(TermDao termDao) {
+        this.termDao = termDao;
+    }
+
+    @Transactional(propagation = Propagation.REQUIRES_NEW)
+    public void removeOrphanedInverseTermRelationships(Term update, Term original) {
+        removeOrphanedRelated(update, original);
+        removeOrphanedRelatedMatch(update, original);
+    }
+
+    private Set<TermInfo> determineOrphaned(Set<TermInfo> newValue, Set<TermInfo> originalValue) {
+        if (originalValue == null || originalValue.isEmpty()) {
+            return Collections.emptySet();
+        }
+        final Set<TermInfo> orphaned = new HashSet<>(originalValue);
```

Comment:
looks like retainall, but you probably need this to avoid NPEs.

Reply:
I don't see it. I need a diff of the two sets - need to keep, which is AFAIK the opposite of `retainAll`. But anyway yes, I also need to avoid NPEs.